projects/sandbox/train/train/validation.py

```diff
@@ -191,32 +196,69 @@ def postprocess(self, preds: torch.Tensor) -> torch.Tensor:
         )
         return preds[0, 0]
 
-    def inject(self, X: torch.Tensor) -> torch.Tensor:
+    def threshold_snrs(self, waveforms: torch.Tensor, psds: torch.Tensor):
+        num_freqs = waveforms.shape[-1] // 2 + 1
+        if num_freqs != psds.shape[-1]:
+            psds = torch.nn.functional.interpolate(psds, (num_freqs,))
+
+        mask = torch.linspace(0, self.sample_rate / 2, num_freqs)
+        mask = mask >= self.highpass
+        mask = mask.to(waveforms.device)
+
+        snrs = gw.compute_network_snr(waveforms, psds, self.sample_rate, mask)
+        target_snrs = snrs.clamp(self.snr_thresh, 1000)
+        weights = target_snrs / snrs
+        return waveforms * weights.view(-1, 1, 1)
+
+    def inject(self, X: torch.Tensor, psds: torch.Tensor) -> torch.Tensor:
+        # downsample the background batch to give us
+        # the desired stride between injections
         X = X[:: self._injection_step]
```

Comment:
I don't understand this step. X has the shape `(batch_size, num_ifos, background_length)`, right? So we're taking every `self._injection_step` kernel?

Reply:
Yeah the idea here is to put some space in between injections (in time) to avoid spurious correlations